File: Solver.hs
module Solver ( Constraint (..), Selector (..), solve, truth, has_n_unique_elements, permutation_reducer, unlist, cross_with) where
    import Data.List

    data Selector = Points [[Int]] | Select [Int -> Bool]
    data Constraint = Matches { predicate :: [Int] -> Bool, selector :: Selector } 
                    | Reducer { reducer :: [[Int]] -> [[Int]], selector :: Selector }

    truth = \_ -> True

    has_n_unique_elements n = (==n) . length . nub  -- True if the list has n unique elements.

    -- A reducer takes a list of options and returns the reduced list of options.
    -- The permutation reducer removes options that would break the permutation constraint.
    permutation_reducer options = reduced_cols where
        bitmapped = map (\x -> map (fromEnum . (`elem` x)) [1..(length options)]) options

        -- Tests if the options in the sequence are the only allowable ones for the set.
        -- [[1,0,0,0]] => True, [[1,1,0,0],[1,1,0,0]] => True
        test_seqs (seq1:seqs) | (sum seq1) == s = all (==seq1) seqs where s = 1 + length seqs
        test_seqs _ = False

        subs = tail $ subsequences [0..(n-1)] where n = length options -- Subsequences of the options to search.

        reduced_cols = foldl' apply_reduction options (filter (test_seqs . (map (bitmapped!!))) subs)

        apply_reduction options seq = zipWith (\i list -> if i `elem` seq then vals else list\\vals) [0..] options
            where vals = options !! (head seq)

    unlist :: Int -> [a] -> [[a]]   -- Turns a flat list into nested lists.
    unlist _ [] = []
    unlist n l = [take n l] ++ (unlist n (drop n l))

    cross_with :: (a -> b -> c) -> [a] -> [b] -> [c]
    cross_with f a b = [f a' b' | a' <- a, b' <- b]

    groupOn f = groupBy (\a b -> f a == f b)

    combos :: [[a]] -> [[a]]  -- Returns the combinations of choosing 1 element from each list.
    combos [] = []
    combos [a] = map (\x -> [x]) a
    combos [a,b] = cross_with (\a b -> [a, b]) a b
    combos (x:xs) = cross_with (:) x (combos xs)

    select :: Selector -> [[Int]] -> [[Int]]  -- Apply selector to points. e.g. select [even] [[x,y]] => [[x,y] : x is even]
    select (Points points) = filter (\p -> any (`isPrefixOf` p) points)
    select (Select filters) = filter (and . (zipWith ($) filters))

    -- Returns all possible solutions. If you want only one solution, use `take 1 $ solve ..`.
    solve :: [Constraint] -> [[Int]] -> [[[Int]]]
    solve constraints field = post_process $ solve' (combos field) where
        post_process results = fmap (transpose . (unlist num_cols)) $ filter (/=[]) $ results

        num_cols = product $ map length (init . init $ field)
        num_values = product $ map length (init field)
        length_constraint = Reducer (\vec -> if length vec >= num_values then vec else []) (Select [])

        solve' solver_step = let
            new_points = apply_constraints (length_constraint:constraints) solver_step
            options = (map . map) last $ groupOn init new_points
            positions = nub $ (map init) new_points

            (Just target) = findIndex (==t_val) lengths
                where lengths = map length options; t_val = head $ filter (>1) lengths          
            with_target = new_points \\ [(positions !! target) ++ [v] | v <- last field, v /= head (options !! target)]
            without_target = new_points \\ [(positions !! target) ++ [head (options !! target)]]

            in ( if all (==1) (map length options) then [concat options] -- One option left for each value or empty list.
                 else (solve' with_target) ++ (solve' without_target) )  -- Split into two branches, push one branch to acc.

    apply_constraints constraints = apply' where
        -- Apply each constraint in sequence, then repeat if there's fewer options until no progress is made.
        apply' points = if length points /= length new_points then apply' new_points else new_points where
            new_points = foldl' (update_points) points constraints
            
            update_points points constraint = let
                selected_points = select (selector $ constraint) points
                options = (map . map) last $ groupOn init selected_points -- [[1,1],[1,2],[2,3]] => [[1,2],[3]]
                positions = nub $ map init selected_points   -- [[1,1],[1,2],[2,3]] => [[1],[2]]
                
                new_options = apply_reducer constraint options positions
                in ( points \\ (selected_points \\ new_options)  )

    -- Applying a `Matches` constraint is reducing by filtering the combinations with the `Matches` predicate.
    apply_reducer (Matches predicate _) options positions = to_point_list . (filter predicate) . combos $ options
        where to_point_list = nub . concat . (map (zipWith (\p v -> p++[v]) positions))

    apply_reducer (Reducer reduce _) options positions = to_point_list . reduce $ options
        where to_point_list = concat . (zipWith (\p -> map (\v -> p++[v])) positions)

File: solver.py
import itertools
from functools import reduce
import operator

# Selector = ('Points', [[Int]]) | ('Select', [fn])
# Constraint = ('Matches', predicate_fn, Selector) | ('Reducer', reducer_fn, Selector)

truth = lambda _: True

def has_n_unique_elements(n): return lambda l: len(list(set(l))) == n

def unlist(n, l): return [l[i:(i+n)] for i in range(0,len(l),n)]
flatten = lambda l: [item for sublist in l for item in sublist]

def nub(l):
    used = []
    return [x for x in l if x not in used and (used.append(x) or True)]

def transpose(l): return list(map(list, zip(*l)))

def list_subtract(a, b):
    if len(b) == 0: return a

    if type(b[0]) == type([]):
        bs = {tuple(item) for item in b}
        return [item for item in a if tuple(item) not in bs]
    else:
        return [item for item in a if item not in set(b)]

def zipWith(f, a, b): return list(map(lambda x: f(*x), zip(a, b)))

def cross_with(f, a, b): return list(map(lambda x: f(*x), itertools.product(a,b)))

def subsequences(l): return flatten([itertools.combinations(l, n) for n in range(1, len(l)+1)])

def permutation_reducer(options):
    def test_seqs(l):
        if sum(l[0]) == 1 + len(l[1:]):
            return all(map(lambda x: x==l[0], l[1:]))
        return False

    bitmapped = list(map(lambda x: list(map(lambda y: y in x, range(1,len(options)+1))), options))
    
    subs = subsequences(range(0,len(options)))
    subs = filter(lambda seqs: test_seqs(list(map(lambda i: bitmapped[i], seqs))), subs)

    for seq in subs:
        vals = options[seq[0]]
        options = list(map(lambda il: il[0] in seq and vals or list_subtract(il[1], vals), enumerate(options)))

    return options

def prod(l): return reduce(operator.mul, l)

def combos(l):
    if l == []:
        return []
    elif len(l) == 1:
        return list(map(lambda x: [x], l[0]))
    elif len(l) == 2:
        return cross_with(lambda a,b: [a,b], l[0], l[1])
    else:
        return cross_with(lambda a,b: [a]+b, l[0], combos(l[1:]))

def isPrefixOf(a,b): return b[:len(a)] == a

def select(selector, l):
    (t, p) = selector
    if t == 'Points':
        return list(filter(lambda x: any(map(lambda y: isPrefixOf(y, x), p)), l))
    else: # t == 'Select'
        return list(filter(lambda x: all(zipWith(lambda f,a: f(a), p, x)), l))


def split_points(points):
    options = []
    positions = []
    for (p, o) in itertools.groupby(points, lambda x: x[:-1]):
        options.append(list(map(lambda x: x[-1], o)))
        positions.append(p)
    return (options, positions)

def solve(constraints, field):
    num_values = prod(map(len, field[:-1]))
    length_constraint = ('Reducer', lambda vec: len(vec) >= num_values and vec or [], ('Select', []))

    def post_process(vals): return transpose(unlist(len(field[1]), vals))

    def solver():
        acc = [combos(field)]

        while True:
            if len(acc) == 0: break

            points = acc.pop()

            points = apply_constraints([length_constraint]+constraints, points)
            (options, positions) = split_points(points)

            if all(map(lambda x: len(x)==1, options)):
                yield post_process(list(map(lambda x: x[0], options)))
            else:
                t_val = next(filter(lambda x: x > 1, map(len, options)))
                target = list(map(len, options)).index(t_val)

                p1 = [positions[target] + [options[target][0]]]
                p2 = [positions[target] + [v] for v in field[-1] if v != options[target][0]]
                without_target = list_subtract(points, p1) # [item for item in points if item not in set(p1)]
                with_target = list_subtract(points, p2) # [item for item in points if item not in set(p2)]

                acc.append(without_target)
                acc.append(with_target)

    return filter(lambda x: x != [], solver())

def apply_constraints(constraints, points):
    def apply(points):
        for c in constraints:
            selected_points = select(c[2], points)
            (options, positions) = split_points(selected_points)

            new_options = apply_reducer(c, options, positions)
            discard = list_subtract(selected_points, new_options) # [item for item in selected_points if item not in set(new_options)]
            points = list_subtract(points, discard) # [item for item in new_points if item not in set(discard)]
        return points

    while True:
        len1 = len(points)
        points = apply(points)
        if len(points) == len1:
            return points

def apply_reducer(constraint, options, positions):
    if constraint[0] == 'Matches':
        a = filter(constraint[1], combos(options))
        return nub(flatten(map(lambda x: zipWith(lambda p,v: p+[v], positions, x), a)))
    else: # == 'Reducer'
        a = constraint[1](options)
        return flatten(zipWith(lambda p,z: list(map(lambda v: p+[v], z)), positions, a))


sudoku01 = "040000179002008054006005008080070910050090030019060040300400700570100200928000060"
def format_sudoku_clues(s): return unlist(9, list(map(int, filter(lambda x: x in "0123456789", s))))

def sudoku_constraints(clues):
    def cell_clue(i,j): return [clues[j-1][i-1]]

    box_selectors = cross_with(lambda a,b: ('Select', [lambda x,a=a: x in a, lambda y,b=b: y in b]), [[1,2,3],[4,5,6],[7,8,9]], [[1,2,3],[4,5,6],[7,8,9]])

    return list(map(lambda s: ('Reducer', permutation_reducer, s), box_selectors)) \
        +[('Matches', lambda vec,x=x,y=y: vec==cell_clue(x,y), ('Points', [[x,y]])) for (x,y) in itertools.product(range(1,10), range(1,10)) if cell_clue(x,y) != [0]] \
        +flatten([[('Reducer', permutation_reducer, ('Select', [lambda x,i=i: x==i, truth])), \
                    ('Reducer', permutation_reducer, ('Select', [truth, lambda y,i=i: y==i]))] for i in range(1,10) ])

def sudoku(clues):
    solution = next(solve(sudoku_constraints(format_sudoku_clues(clues)), [range(1,10),range(1,10),range(1,10)]))
    # show_solution [] = putStrLn "No solution found"
    # show_solution [solution] = show_sudoku_solution solution

    def show_sudoku_solution(grid): 
        dash_line = '-'*21 + "\n"

        def print_line(line):
            if line == []: return "----"
            [c1,c2,c3] = unlist(3, line)
            return ' '.join(map(str, c1)) + ' | ' + ' '.join(map(str, c2)) + ' | ' + ' '.join(map(str, c3)) + '\n'

        [r1, r2, r3] = unlist(3, grid)
        grid_string = ''.join(map(print_line, r1)) + dash_line + ''.join(map(print_line, r2)) + dash_line + ''.join(map(print_line, r3))
        
        print(grid_string)
        
    show_sudoku_solution(solution)
